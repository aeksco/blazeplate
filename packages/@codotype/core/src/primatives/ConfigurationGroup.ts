import { ConfigurationGroup } from "..";
import {
    GroupLayoutVariant,
    ConfigurationGroupSection,
    ConfigurationGroupProperty,
} from "../configuration-option-types";

interface ConfigurationGroupBuilderParams {
    label: string;
    identifier: string;
    description: string;
    documentation?: string;
    enabled?: boolean;
    allowDisable?: boolean;
    layoutVariant?: GroupLayoutVariant;
    sections?: ConfigurationGroupSection[];
    properties?: ConfigurationGroupProperty[];
}

export class ConfigurationGroupBuilder implements ConfigurationGroup {
    label: string;
    identifier: string;
    description: string;
    layoutVariant: GroupLayoutVariant = GroupLayoutVariant.LIST;
    sections: ConfigurationGroupSection[] = [];
    properties: ConfigurationGroupProperty[] = [];
    documentation: string = "";
    enabledByDefault: boolean = true;
    allowDisable: boolean = false;

    constructor(params: ConfigurationGroupBuilderParams) {
        this.label = params.label;
        this.identifier = params.identifier;
        this.description = params.description;

        this.layoutVariant = params.layoutVariant || this.layoutVariant;
        this.sections = params.sections || this.sections;
        this.properties = params.properties || this.properties;
        this.documentation = params.documentation || this.documentation;
        this.enabledByDefault =
            params.enabled !== undefined
                ? params.enabled
                : this.enabledByDefault;
        this.allowDisable =
            params.allowDisable !== undefined
                ? params.allowDisable
                : this.allowDisable;
    }
}
