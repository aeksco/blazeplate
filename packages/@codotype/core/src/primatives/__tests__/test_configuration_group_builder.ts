import { ConfigurationGroupBuilder } from "../ConfigurationGroup";

// // // //

describe("factory", () => {
    test("works", () => {
        const configurationGroup = new ConfigurationGroupBuilder({
            label: "API Examples",
            identifier: "api_examples",
            description: "Turn API Examples on/off",
        });

        expect(configurationGroup).toMatchSnapshot();
    });
});
