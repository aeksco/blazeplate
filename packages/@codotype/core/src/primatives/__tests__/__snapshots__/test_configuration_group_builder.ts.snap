// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`factory works 1`] = `
ConfigurationGroupBuilder {
  "allowDisable": false,
  "description": "Turn API Examples on/off",
  "documentation": "",
  "enabledByDefault": true,
  "identifier": "api_examples",
  "label": "API Examples",
  "layoutVariant": "LIST",
  "properties": Array [],
  "sections": Array [],
}
`;
